File: scripts/bindings/ggplot_legends.py
"""FIX-GG01: execute the Rust matrix inputs through the actual primary Python adapter.

Arguments: built Python module directory, Rust artifact directory, output directory.
The Rust test supplies independent legend-count/color/provenance expectations.
"""
import json
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
sys.path[:0] = [str(ROOT / 'packages/python'), str(Path(sys.argv[1]).resolve())]
import finstack_chart as c

inputs, out = map(Path, sys.argv[2:4])
out.mkdir(parents=True, exist_ok=True)
output = c.Output((ROOT / 'fixtures/capability/fonts/NotoSans-Regular.ttf').read_bytes())
files = sorted(inputs.glob('*.plot.json'))
assert len(files) == 24, len(files)
for file in files:
    name = file.name.removesuffix('.plot.json')
    plot = c.Plot.from_json(file.read_text())
    chart = plot.chart()
    if name.startswith('hidden-'):
        chart.legend_visible(False)
    width, height = (130, 65) if name.startswith('tight-') else (500, 300)
    options = c.export_options(width, height).dpi(96).basis('current')
    request = chart.request(output, options)
    frame = request.prepare()
    scene = frame.scene()
    expected = json.loads((inputs / f'{name}.scene.json').read_text())
    assert scene == expected, name
    (out / f'{name}.scene.json').write_text(json.dumps(scene, indent=2))
    for fmt in ('svg', 'pdf', 'png'):
        (out / f'{name}.{fmt}').write_bytes(frame.export(fmt))
    # Exercise host component dispatch as well as the primary interchange path.
    if name.startswith('two-entry-'):
        edited = plot.edit().legend(c.legend().scale('series').untitled()).build()
        chart.apply_plot(edited, 0)
        edited_request = chart.request(output, options)
        edited_frame = edited_request.prepare()
        texts = [i['primitive']['Text']['text'] for i in edited_frame.scene()['items'] if 'Text' in i['primitive']]
        assert 'Series' not in texts and 'Color' not in texts and texts.count('Alpha') == 1
        edited_frame.dispose(); edited_request.dispose(); edited.dispose()
        generic = plot.edit().legend(c.legend().scale('series').generic_title()).build()
        generic_request = output.request(generic, options)
        generic_frame = generic_request.prepare()
        texts = [i['primitive']['Text']['text'] for i in generic_frame.scene()['items'] if 'Text' in i['primitive']]
        assert texts.count('Color') == 1 and 'Series' not in texts and texts.count('Alpha') == 1
        generic_frame.dispose(); generic_request.dispose(); generic.dispose()
    frame.dispose(); request.dispose(); chart.dispose(); plot.dispose(); options.dispose()
output.dispose()
print('PASS FIX-GG01 Python: 24 exact scenes, 72 exports and 6 host untitled/generic edits.')

File: crates/chart-export/tests/ggplot_legends.rs
//! FIX-GG01: shared legend acceptance through primary authoring and real publication.
use chart_core::{prelude::*, scene::Primitive, state::ChartAction, theme::rgb};
use chart_export::*;

fn authored(case: &str, faceted: bool, collect: bool) -> Plot {
    let data = Data::columns()
        .column("x", [1., 2.])
        .column("y", [2., 4.])
        .column("series", ["Alpha", "Beta"])
        .column("panel", ["One", "One"])
        .build()
        .unwrap();
    let mut scale = color_discrete("series").palette(vec![rgb(0, 80, 180), rgb(200, 20, 40)]);
    if case == "empty" {
        scale = scale.domain(Vec::<String>::new());
    }
    let mut builder = plot(data)
        .aes(aes().x("x").y("y").color("series"))
        .scale(scale)
        .layer(points())
        .legend(if case == "untitled" {
            legend().scale("series").untitled()
        } else {
            legend().scale("series").title("Series")
        });
    if case == "shared" {
        builder = builder.layer(points());
    } else if case == "incompatible" {
        builder = builder
            .scale(color_discrete("other").palette(vec![rgb(200, 20, 40), rgb(0, 80, 180)]))
            .layer(points().aes(aes().color("series").color_scale("other")))
            .legend(legend().scale("other").title("Other"));
    }
    if faceted {
        builder = builder.facet(facet_wrap("panel").collect_guides(collect));
    }
    let plot = builder.build().unwrap();
    if case == "edited-untitled" {
        plot.edit()
            .legend(legend().scale("series").untitled())
            .build()
            .unwrap()
    } else {
        plot
    }
}

fn text(item: &chart_core::scene::SceneItem) -> Option<&str> {
    match &item.primitive {
        Primitive::Text { text, .. } => Some(text),
        Primitive::GlyphRun { run, .. } => Some(&run.text),
        _ => None,
    }
}

#[test]
fn fix_gg01_primary_legend_matrix() {
    let output = Output::new(
        include_bytes!("../../../fixtures/capability/fonts/NotoSans-Regular.ttf").as_slice(),
    )
    .unwrap();
    for (faceted, collect) in [(false, true), (true, true), (true, false)] {
        for case in [
            "two-entry",
            "empty",
            "hidden",
            "tight",
            "shared",
            "incompatible",
            "untitled",
            "edited-untitled",
        ] {
            let name = format!("{case}-facet{faceted}-collect{collect}");
            let plot =
                Plot::from_json(&authored(case, faceted, collect).to_json().unwrap()).unwrap();
            let mut chart = plot.chart().unwrap();
            if case == "hidden" {
                chart.act(ChartAction::SetLegendVisible(false)).unwrap();
            }
            let (width, height) = if case == "tight" {
                (130., 65.)
            } else {
                (500., 300.)
            };
            let options = export_options(PageSize::points(width, height).unwrap())
                .dpi(96)
                .basis(CaptureBasis::Current);
            let snapshot = output
                .live_request(&chart, options)
                .unwrap()
                .prepare()
                .unwrap();
            let scene = snapshot.layout().scene();
            let count = |expected| {
                scene
                    .items()
                    .iter()
                    .filter(|i| text(i) == Some(expected))
                    .count()
            };
            let visible = !matches!(case, "hidden" | "empty");
            if case != "tight" {
                assert_eq!(
                    count("Series"),
                    usize::from(visible && !case.ends_with("untitled")),
                    "{name}"
                );
                assert_eq!(
                    count("Color"),
                    0,
                    "{name}: an omitted title must stay omitted"
                );
                let copies = if case == "incompatible" {
                    2
                } else {
                    usize::from(visible)
                };
                assert_eq!(count("Alpha"), copies, "{name}");
                assert_eq!(count("Beta"), copies, "{name}");
                assert_eq!(
                    count("Other"),
                    usize::from(case == "incompatible"),
                    "{name}"
                );
            } else {
                assert!(
                    snapshot.layout().diagnostics().iter().any(|d| {
                        d.code == DiagnosticCode::LayoutPressure
                            && d.message.contains("Legend pressure")
                    }),
                    "{name}: clipping/omission must be reported"
                );
            }
            let blue = rgb(0, 80, 180);
            let red = rgb(200, 20, 40);
            let swatches: Vec<_> = scene
                .items()
                .iter()
                .enumerate()
                .filter_map(|(index, i)| match i.primitive {
                    Primitive::Rectangle { bounds, fill }
                        if i.layer.is_none() && [blue, red].contains(&fill) =>
                    {
                        assert!(
                            snapshot.layout().targets()[index].is_empty(),
                            "{name}: guide acquired a source target"
                        );
                        let clip = i.clip.expect("guide clip");
                        assert!(
                            bounds.origin().x() >= clip.origin().x()
                                && bounds.max_x() <= clip.max_x()
                        );
                        assert!(
                            bounds.origin().y() >= clip.origin().y()
                                && bounds.max_y() <= clip.max_y()
                        );
                        Some(fill)
                    }
                    _ => None,
                })
                .collect();
            if case != "tight" {
                let expected = if case == "incompatible" {
                    vec![blue, red, red, blue]
                } else if visible {
                    vec![blue, red]
                } else {
                    vec![]
                };
                assert_eq!(
                    swatches, expected,
                    "{name}: shared guides deduplicate; incompatible palettes remain separate"
                );
                let points: Vec<_> = scene
                    .items()
                    .iter()
                    .enumerate()
                    .filter(|(_, i)| {
                        i.layer.is_some() && matches!(i.primitive, Primitive::Point { .. })
                    })
                    .collect();
                assert_eq!(
                    points.len(),
                    if matches!(case, "shared" | "incompatible") {
                        4
                    } else {
                        2
                    },
                    "{name}"
                );
                for (index, _) in points {
                    assert!(
                        !snapshot.layout().targets()[index].is_empty(),
                        "{name}: source provenance missing"
                    );
                }
            }
            assert_eq!(scene.items().len(), snapshot.layout().targets().len());
            // Set explicitly when retaining the same tested inputs for host and visual evidence.
            let directory = std::env::var_os("GG_LEGEND_ARTIFACTS").map(std::path::PathBuf::from);
            if let Some(directory) = &directory {
                std::fs::create_dir_all(directory).unwrap();
                std::fs::write(
                    directory.join(format!("{name}.plot.json")),
                    plot.to_json().unwrap(),
                )
                .unwrap();
                std::fs::write(
                    directory.join(format!("{name}.scene.json")),
                    snapshot.scene_json().unwrap(),
                )
                .unwrap();
            }
            for (format, extension) in [
                (Format::Svg, "svg"),
                (Format::Pdf, "pdf"),
                (Format::Png, "png"),
            ] {
                let artifact = snapshot.export(format).unwrap();
                match format {
                    Format::Svg => assert!(
                        std::str::from_utf8(&artifact.bytes)
                            .unwrap()
                            .contains("<svg")
                    ),
                    Format::Pdf => assert!(artifact.bytes.starts_with(b"%PDF-")),
                    Format::Png => {
                        let mut reader = png::Decoder::new(artifact.bytes.as_slice())
                            .read_info()
                            .unwrap();
                        let mut pixels = vec![0; reader.output_buffer_size()];
                        reader.next_frame(&mut pixels).unwrap();
                    }
                    _ => unreachable!("This legend matrix enumerates SVG, PDF and PNG above."),
                }
                if let Some(directory) = &directory {
                    artifact
                        .save(directory.join(format!("{name}.{extension}")))
                        .unwrap();
                }
            }
        }
    }
}
